src/java/org/apache/cassandra/index/sai/disk/oldlucene/LegacyByteBuffersDataInput.java

```diff
@@ -0,0 +1,359 @@
+     * enough remaining limit.
+     *
+     * If there are fewer than {@code len} bytes in the input, {@link EOFException}
+     * is thrown.
+     */
+    public void readBytes(ByteBuffer buffer, int len) throws EOFException {
+        try {
+            while (len > 0) {
+                ByteBuffer block = blocks[blockIndex(pos)].duplicate();
+                int blockOffset = blockOffset(pos);
+                block.position(blockOffset);
+                int chunk = Math.min(len, block.remaining());
+                if (chunk == 0) {
+                    throw new EOFException();
+                }
+
+                // Update pos early on for EOF detection on output buffer, then try to get buffer content.
+                pos += chunk;
+                block.limit(blockOffset + chunk);
+                buffer.put(block);
+
+                len -= chunk;
+            }
+        } catch (BufferUnderflowException | ArrayIndexOutOfBoundsException e) {
+            if (pos >= size()) {
+                throw new EOFException();
+            } else {
+                throw e; // Something is wrong.
+            }
+        }
+    }
+
+    @Override
+    public void readBytes(byte[] arr, int off, int len) throws EOFException {
+        try {
+            while (len > 0) {
+                ByteBuffer block = blocks[blockIndex(pos)].duplicate();
+                block.position(blockOffset(pos));
+                int chunk = Math.min(len, block.remaining());
+                if (chunk == 0) {
+                    throw new EOFException();
+                }
+
+                // Update pos early on for EOF detection, then try to get buffer content.
+                pos += chunk;
+                block.get(arr, off, chunk);
+
+                len -= chunk;
+                off += chunk;
+            }
+        } catch (BufferUnderflowException | ArrayIndexOutOfBoundsException e) {
+            if (pos >= size()) {
+                throw new EOFException();
+            } else {
+                throw e; // Something is wrong.
+            }
+        }
+    }
+
+    @Override
+    public short readShort() throws IOException
+    {
+        return (short) (((readByte() & 0xFF) <<  8) |  (readByte() & 0xFF));
+    }
+
+    @Override
+    public int readInt() throws IOException
+    {
+        return ((readByte() & 0xFF) << 24) | ((readByte() & 0xFF) << 16)
+               | ((readByte() & 0xFF) << 8) | (readByte() & 0xFF);
+    }
+
+    @Override
+    public long readLong() throws IOException
+    {
+        return (((long)readInt()) << 32) | (readInt() & 0xFFFFFFFFL);
+    }
+
+    @Override
+    public void skipBytes(long l) throws IOException
+    {
+        if (l < 0) {
+            throw new IllegalArgumentException("l must be >= 0, got " + l);
+        }
+        if (l > size() - pos) {
+            throw new EOFException();
+        }
+        pos += l;
+    }
+
+    @Override
+    public byte readByte(long pos) {
```

Comment:
issue: Shouldn't it assert that pos is within the valid range? I can see it might fail with AIOOBE if pos is incorrect.